### PR TITLE
Adding command line options

### PR DESCRIPTION
**Pull request by MaxOstrowski**
Open PR, nothing done yet.

Add options for:
 - refine reasons
 - refine introduce
 - sort elements
 - propagate prev lit (may happen on a per thread basis)
 - magic clause (definitely needs a better name)
 - literals only
 - magic weight constraint

### CODE REVIEW

**Comment by rkaminsk** [comment]
I made a first pass over the pull request. There are still some things I want to change.
- I do not like the duplication of members in the `Config` and `AppConfig` objects. I'll think a bit more if there is a nicer way.
- We could probably also make the `check_state` and `check_solution` variables options. Why not have debug options?
- The option descriptions should be refined further. The more similar they look to clasp's descriptions the better.
- Btw. clasp has a solution for on/off flags. It uses `--do` and the corresponding flag `--no-do`. Depending on the current defaults on of them has no effect.

File: examples/clingo/csp/csp/__main__.py [hunk 1]
```diff
@@ -31,43 +52,113 @@ def print_model(self, model, default_printer):
...
+                    self._parse_bool_thread(self.config, "prev_lit"), True, argument="True/False[,i]")
+        options.add(group, "shift-constraints",
+                    "Shift constraints into head in integrity constraints [{}]".format(self.config.shift_constraints),
+                    self._parse_bool(self.config, "shift_constraints"))
+        options.add(group, "check-solution@2",  # TODO: @ notation does not seem to work
+                    "Debug option to verify solutions [{}]".format(conf.check_solution),
+                    self._parse_bool(conf, "check_solution"))
+        options.add(group, "check-state@2",
+                    "Debug option to verify state consistency [{}]".format(conf.check_state),
+                    self._parse_bool(conf, "check_state"))
```
Comment (MaxOstrowski) [inline examples/clingo/csp/csp/__main__.py]: This level mechanism does not seem to work or I got it wrong

**Comment by rkaminsk** [comment]
The syntax should be "name,@level".

File: examples/clingo/csp/csp/__main__.py (hunk 1 above)
Comment (rkaminsk) [reply examples/clingo/csp/csp/__main__.py]: The syntax should be "name,@Level".

Comment (rkaminsk) [reply examples/clingo/csp/csp/__main__.py]: I should probably update the docstring to also include the comma.

Comment (MaxOstrowski) [reply examples/clingo/csp/csp/__main__.py]: thx

**Comment by MaxOstrowski** [comment]
Thanks a lot, anything I can do ?

> I do not like the duplication of members in the Config and AppConfig objects. I'll think a bit more if there is a nicer way.

That's the only thing left from your last review...

**Comment by rkaminsk** [comment]
You can have a look. I rearranged a bit and also pushed an update clingo version that allows for using `clingo.Flag` like a `bool`. If you are fine with the changes, we can merge.

File: examples/clingo/csp/csp/__main__.py [hunk 2]
```diff
@@ -42,18 +31,21 @@ def __init__(self):
...
 
-        print("Valid assignment for constraints found:")
-        print(" ".join("{}={}".format(n, v) for n, v in ass))
-
+        sys.stdout.write("Valid assignment for constraints found:\n")
+        sep = ""
+        for n, v in self._propagator.get_assignment(model.thread_id):
+            sys.stdout.write("{}{}={}".format(sep, n, v))
+            sep = " "
+        sys.stdout.write("\n")
```
Comment (MaxOstrowski) [inline examples/clingo/csp/csp/__main__.py]: Any reason for not using print in this commit ?

Comment (rkaminsk) [reply examples/clingo/csp/csp/__main__.py]: In principle, the print function could be used. But the differences between python 2 and 3 make this annoying. Also, I do not need any of it's extra features. In the end it's just a wrapper around sys.stdout anyway.

**Comment by rkaminsk** [comment]
Btw. there are still some pylint warnings which I am going to address later. I only have a machine with an older pylint here which I am not going to bother with.

**Pull request merged**